File: src/vdbfusion/vdbfusion/VDBVolume.cpp
// MIT License
//
// # Copyright (c) 2022 Ignacio Vizzo, Cyrill Stachniss, University of Bonn
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

#include "VDBVolume.h"

// OpenVDB
#include <openvdb/Types.h>
#include <openvdb/math/BBox.h>
#include <openvdb/math/DDA.h>
#include <openvdb/math/Ray.h>
#include <openvdb/openvdb.h>
#include <openvdb/tools/Clip.h>
#include <openvdb/tools/GridOperators.h>

#include <Eigen/Core>
#include <algorithm>
#include <cmath>
#include <functional>
#include <iostream>
#include <memory>
#include <tuple>
#include <vector>

#include "utils/conversions.h"

using Sophus::SE3d;
using PointCloud = std::vector<Eigen::Vector3d>;

typedef Eigen::Matrix<double, 3, 6> Matrix3x6d;
typedef Eigen::Matrix<double, 6, 6> Matrix6x6d;
typedef Eigen::Matrix<double, 6, 1> Matrix6x1d;

namespace {
float ComputeSDF(const Eigen::Vector3d& origin,
                 const Eigen::Vector3d& point,
                 const Eigen::Vector3d& voxel_center) {
    const Eigen::Vector3d v_voxel_origin = voxel_center - origin;
    const Eigen::Vector3d v_point_voxel = point - voxel_center;
    const double dist = v_point_voxel.norm();
    const double proj = v_voxel_origin.dot(v_point_voxel);
    const double sign = proj / std::abs(proj);
    return static_cast<float>(sign * dist);
}

Eigen::Vector3d GetVoxelCenter(const openvdb::Coord& voxel, const openvdb::math::Transform& xform) {
    const float voxel_size = xform.voxelSize()[0];
    openvdb::math::Vec3d v_wf = xform.indexToWorld(voxel) + voxel_size / 2.0;
    return Eigen::Vector3d(v_wf.x(), v_wf.y(), v_wf.z());
}

PointCloud ApplyTransform(const PointCloud& pcl, const SE3d& T) {
    auto R = T.rotationMatrix();
    auto tr = T.translation();

    PointCloud pcl_t;
    pcl_t.reserve(pcl.size());

    std::for_each(pcl.cbegin(), pcl.cend(),
                  [&](const Eigen::Vector3d& pt) { pcl_t.emplace_back((R * pt) + tr); });
    return pcl_t;
}

}  // namespace

namespace vdbfusion {

VDBVolume::VDBVolume(float voxel_size, float sdf_trunc, bool space_carving /* = false*/)
    : voxel_size_(voxel_size), sdf_trunc_(sdf_trunc), space_carving_(space_carving) {
    tsdf_ = openvdb::FloatGrid::create(sdf_trunc_);
    tsdf_->setName("D(x): signed distance grid");
    tsdf_->setTransform(openvdb::math::Transform::createLinearTransform(voxel_size_));
    tsdf_->setGridClass(openvdb::GRID_LEVEL_SET);

    weights_ = openvdb::FloatGrid::create(0.0f);
    weights_->setName("W(x): weights grid");
    weights_->setTransform(openvdb::math::Transform::createLinearTransform(voxel_size_));
    weights_->setGridClass(openvdb::GRID_UNKNOWN);
}

void VDBVolume::UpdateTSDF(const float& sdf,
                           const openvdb::Coord& voxel,
                           const std::function<float(float)>& weighting_function) {
    using AccessorRW = openvdb::tree::ValueAccessorRW<openvdb::FloatTree>;
    if (sdf > -sdf_trunc_) {
        AccessorRW tsdf_acc = AccessorRW(tsdf_->tree());
        AccessorRW weights_acc = AccessorRW(weights_->tree());
        const float tsdf = std::min(sdf_trunc_, sdf);
        const float weight = weighting_function(sdf);
        const float last_weight = weights_acc.getValue(voxel);
        const float last_tsdf = tsdf_acc.getValue(voxel);
        const float new_weight = weight + last_weight;
        const float new_tsdf = (last_tsdf * last_weight + tsdf * weight) / (new_weight);
        tsdf_acc.setValue(voxel, new_tsdf);
        weights_acc.setValue(voxel, new_weight);
    }
}

void VDBVolume::Integrate(openvdb::FloatGrid::Ptr grid,
                          const std::function<float(float)>& weighting_function) {
    for (auto iter = grid->cbeginValueOn(); iter.test(); ++iter) {
        const auto& sdf = iter.getValue();
        const auto& voxel = iter.getCoord();
        this->UpdateTSDF(sdf, voxel, weighting_function);
    }
}

void VDBVolume::Integrate(const PointCloud& points,
                          const Eigen::Vector3d& origin,
                          const std::function<float(float)>& weighting_function) {
    if (points.empty()) {
        std::cerr << "PointCloud provided is empty\n";
        return;
    }

    // Get some variables that are common to all rays
    const openvdb::math::Transform& xform = tsdf_->transform();
    const openvdb::Vec3R eye(origin.x(), origin.y(), origin.z());

    // Get the "unsafe" version of the grid acessors
    auto tsdf_acc = tsdf_->getUnsafeAccessor();
    auto weights_acc = weights_->getUnsafeAccessor();

    // Launch an for_each execution, use std::execution::par to parallelize this region
    std::for_each(points.cbegin(), points.cend(), [&](const auto& point) {
        // Get the direction from the sensor origin to the point and normalize it
        const Eigen::Vector3d direction = point - origin;
        openvdb::Vec3R dir(direction.x(), direction.y(), direction.z());
        dir.normalize();

        // Truncate the Ray before and after the source unless space_carving_ is specified.
        const auto depth = static_cast<float>(direction.norm());
        const float t0 = space_carving_ ? 0.0f : depth - sdf_trunc_;
        const float t1 = depth + sdf_trunc_;

        // Create one DDA per ray(per thread), the ray must operate on voxel grid coordinates.
        const auto ray = openvdb::math::Ray<float>(eye, dir, t0, t1).worldToIndex(*tsdf_);
        openvdb::math::DDA<decltype(ray)> dda(ray);
        do {
            const auto voxel = dda.voxel();
            const auto voxel_center = GetVoxelCenter(voxel, xform);
            const auto sdf = ComputeSDF(origin, point, voxel_center);
            if (sdf > -sdf_trunc_) {
                const float tsdf = std::min(sdf_trunc_, sdf);
                const float weight = weighting_function(sdf);
                const float last_weight = weights_acc.getValue(voxel);
                const float last_tsdf = tsdf_acc.getValue(voxel);
                const float new_weight = weight + last_weight;
                const float new_tsdf = (last_tsdf * last_weight + tsdf * weight) / (new_weight);
                tsdf_acc.setValue(voxel, new_tsdf);
                weights_acc.setValue(voxel, new_weight);
            }
        } while (dda.step());
    });
}

openvdb::FloatGrid::Ptr VDBVolume::Prune(float min_weight) const {
    const auto weights = weights_->tree();
    const auto tsdf = tsdf_->tree();
    const auto background = sdf_trunc_;
    openvdb::FloatGrid::Ptr clean_tsdf = openvdb::FloatGrid::create(sdf_trunc_);
    clean_tsdf->setName("D(x): Pruned signed distance grid");
    clean_tsdf->setTransform(openvdb::math::Transform::createLinearTransform(voxel_size_));
    clean_tsdf->setGridClass(openvdb::GRID_LEVEL_SET);
    clean_tsdf->tree().combine2Extended(tsdf, weights, [=](openvdb::CombineArgs<float>& args) {
        if (args.aIsActive() && args.b() > min_weight) {
            args.setResult(args.a());
            args.setResultIsActive(true);
        } else {
            args.setResult(background);
            args.setResultIsActive(false);
        }
    });
    return clean_tsdf;
}

ImplicitRegistration::ImplicitRegistration(VDBVolume& vdb_volume_global,
                                           const registrationConfigParams& config)
    : vdb_volume_global_(vdb_volume_global), config_(config) {}

openvdb::tools::ScalarToVectorConverter<openvdb::FloatGrid>::Type::Ptr
ImplicitRegistration::ComputeGradient(const openvdb::FloatGrid::Ptr grid) const {
    openvdb::tools::Gradient<openvdb::FloatGrid,
                             openvdb::tools::gridop::ToMaskGrid<openvdb::FloatGrid>::Type,
                             openvdb::util::NullInterrupter, openvdb::math::CD_4TH>
        grad_op(*grid);
    return grad_op.process(true);
}

openvdb::FloatGrid::Ptr ImplicitRegistration::ClipVolume(const SE3d& T) const {
    auto center = Eigen2VDB::Vec3(T.translation());
    auto clip_val = config_.clipping_range_;
    openvdb::BBoxd bbox(center - clip_val, center + clip_val);

    auto R = T.rotationMatrix();

    bbox.expand(center +
                Eigen2VDB::Vec3<double>(R * Eigen::Vector3d{-clip_val, -clip_val, -clip_val}));
    bbox.expand(center +
                Eigen2VDB::Vec3<double>(R * Eigen::Vector3d{-clip_val, -clip_val, clip_val}));
    bbox.expand(center +
                Eigen2VDB::Vec3<double>(R * Eigen::Vector3d{-clip_val, clip_val, -clip_val}));
    bbox.expand(center +
                Eigen2VDB::Vec3<double>(R * Eigen::Vector3d{-clip_val, clip_val, clip_val}));
    bbox.expand(center +
                Eigen2VDB::Vec3<double>(R * Eigen::Vector3d{clip_val, -clip_val, -clip_val}));
    bbox.expand(center +
                Eigen2VDB::Vec3<double>(R * Eigen::Vector3d{clip_val, -clip_val, clip_val}));
    bbox.expand(center +
                Eigen2VDB::Vec3<double>(R * Eigen::Vector3d{clip_val, clip_val, -clip_val}));
    bbox.expand(center +
                Eigen2VDB::Vec3<double>(R * Eigen::Vector3d{clip_val, clip_val, clip_val}));
    return openvdb::tools::clip(*vdb_volume_global_.tsdf_, bbox);
}

SE3d ImplicitRegistration::ConstantVelocityModel() const {
    SE3d T_pred{};
    T_pred.rotationMatrix() = T_2.rotationMatrix().transpose() * T_1.rotationMatrix();
    T_pred.translation() =
        T_2.rotationMatrix().transpose() * (T_1.translation() - T_2.translation());
    return T_pred;
}

std::tuple<PointCloud, SE3d, int> ImplicitRegistration::AlignScan(const PointCloud& points,
                                                                  const SE3d& init_tf) {
    // const openvdb::FloatGrid::Ptr local_tsdf = this->ClipVolume(init_tf);
    auto local_tsdf = vdb_volume_global_.tsdf_;
    const auto local_tsdf_grad = this->ComputeGradient(local_tsdf);
    const auto xform = local_tsdf->transform();

    const auto tsdf_acc = local_tsdf->getConstUnsafeAccessor();
    const auto grad_acc = local_tsdf_grad->getConstUnsafeAccessor();

    auto T = init_tf;
    if (config_.use_constant_velocity_model_) T = T * ConstantVelocityModel();

    Matrix6x6d A;
    Matrix6x1d b;

    // clang-format off
    Matrix3x6d grad_tf;
    grad_tf << 1, 0, 0, 0, 0, 0,
               0, 1, 0, 0, 0, 0,
               0, 0, 1, 0, 0, 0;
    // clang-format on

    int n_iters = 0;
    auto m_scale = vdb_volume_global_.sdf_trunc_ / 3;

    while (true) {
        A.setConstant(0);
        b.setConstant(0);

        auto points_g = ApplyTransform(points, T);

        std::for_each(points_g.cbegin(), points_g.cend(), [&](const Eigen::Vector3d& point_g) {
            auto index_pos = xform.worldToIndex(Eigen2VDB::Vec3(point_g));
            auto voxel = openvdb::math::Coord(index_pos.x(), index_pos.y(), index_pos.z());
            float distance = tsdf_acc.getValue(voxel);

            // Sensitive to this params
            if (std::abs(distance) < vdb_volume_global_.sdf_trunc_) {
                auto grad_sdf = VDB2Eigen::Vec3<double>(grad_acc.getValue(voxel));

                grad_tf.block<3, 3>(0, 3) = -1 * Sophus::SO3d::hat(point_g);
                auto grad = grad_tf.transpose() * grad_sdf;

                double weight = m_scale / std::pow(std::pow(distance, 2) + m_scale, 2);
                A += weight * grad * grad.transpose();
                b -= weight * distance * grad;
            }
        });
        // auto se3_old = T.log();
        auto dx = A.ldlt().solve(b);
        auto dT = SE3d::exp(dx);
        T = dT * T;

        T_2 = T_1;
        T_1 = T;

        n_iters++;

        if (n_iters > config_.max_iters_ || dx.lpNorm<2>() < config_.convergence_threshold_) {
            std::cout << "Scan Aligning converged; n_iters = " << n_iters << "\n";
            break;
        }
    }
    auto aligned_points = ApplyTransform(points, T);
    return std::make_tuple(aligned_points, T, n_iters);
}

void ImplicitRegistration::RMSError(const PointCloud& points) {
    double sum = 0;
    auto tsdf_acc = vdb_volume_global_.tsdf_->getConstUnsafeAccessor();
    auto local_tsdf_xform = vdb_volume_global_.tsdf_->transform();
    std::for_each(points.cbegin(), points.cend(), [&](const Eigen::Vector3d& point) {
        auto voxel_tsdf =
            openvdb::math::Coord::round(local_tsdf_xform.worldToIndex(Eigen2VDB::Vec3(point)));
        sum += std::pow(tsdf_acc.getValue(voxel_tsdf), 2);
    });
    std::cout << "RMS: " << std::sqrt(sum / points.size()) << "\n";
}
}  // namespace vdbfusion

File: src/vdbfusion/pybind/vdb_volume.py
# MIT License
#
# # Copyright (c) 2022 Ignacio Vizzo, Cyrill Stachniss, University of Bonn
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

from typing import Any, Optional, Tuple, Callable, overload

import numpy as np

from . import vdbfusion_pybind


class VDBVolume:
    def __init__(
        self,
        voxel_size: float,
        sdf_trunc: float,
        space_carving: bool = False,
    ):
        self._volume = vdbfusion_pybind._VDBVolume(
            voxel_size=np.float32(voxel_size),
            sdf_trunc=np.float32(sdf_trunc),
            space_carving=space_carving,
        )
        # Passthrough all data members from the C++ API
        self.voxel_size = self._volume._voxel_size
        self.sdf_trunc = self._volume._sdf_trunc
        self.space_carving = self._volume._space_carving
        self.pyopenvdb_support_enabled = self._volume.PYOPENVDB_SUPPORT_ENABLED
        if self.pyopenvdb_support_enabled:
            self.tsdf = self._volume._tsdf
            self.weights = self._volume._weights

    def __repr__(self) -> str:
        return (
            f"VDBVolume with:\n"
            f"voxel_size    = {self.voxel_size}\n"
            f"sdf_trunc     = {self.sdf_trunc}\n"
            f"space_carving = {self.space_carving}\n"
        )

    @overload
    def integrate(
        self,
        points: np.ndarray,
        extrinsic: np.ndarray,
        weighting_function: Callable[[float], float],
    ) -> None:
        ...

    @overload
    def integrate(self, points: np.ndarray, extrinsic: np.ndarray, weight: float) -> None:
        ...

    @overload
    def integrate(self, points: np.ndarray, extrinsic: np.ndarray) -> None:
        ...

    @overload
    def integrate(self, grid, weighting_function: Callable[[float], float]) -> None:
        ...

    @overload
    def integrate(self, grid, weight: float) -> None:
        ...

    @overload
    def integrate(self, grid) -> None:
        ...

    def integrate(
        self,
        points: Optional[np.ndarray] = None,
        extrinsic: Optional[np.ndarray] = None,
        grid: Optional[Any] = None,
        weight: Optional[float] = None,
        weighting_function: Optional[Callable[[float], float]] = None,
    ) -> None:
        if grid is not None:
            if not self.pyopenvdb_support_enabled:
                raise NotImplementedError("Please compile with PYOPENVDB_SUPPORT_ENABLED")
            if weighting_function is not None:
                return self._volume._integrate(grid, weighting_function)
            if weight is not None:
                return self._volume._integrate(grid, weight)
            return self._volume._integrate(grid)
        else:
            assert isinstance(points, np.ndarray), "points must by np.ndarray(n, 3)"
            assert points.dtype == np.float64, "points dtype must be np.float64"
            assert isinstance(extrinsic, np.ndarray), "origin/extrinsic must by np.ndarray"
            assert extrinsic.dtype == np.float64, "origin/extrinsic dtype must be np.float64"
            assert extrinsic.shape in [
                (3,),
                (3, 1),
                (4, 4),
            ], "origin/extrinsic must be a (3,) array or a (4,4) matrix"

            _points = vdbfusion_pybind._VectorEigen3d(points)
            if weighting_function is not None:
                return self._volume._integrate(_points, extrinsic, weighting_function)
            if weight is not None:
                return self._volume._integrate(_points, extrinsic, weight)
            self._volume._integrate(_points, extrinsic)

    @overload
    def update_tsdf(
        self, sdf: float, ijk: np.ndarray, weighting_function: Optional[Callable[[float], float]]
    ) -> None:
        ...

    @overload
    def update_tsdf(self, sdf: float, ijk: np.ndarray) -> None:
        ...

    def update_tsdf(
        self,
        sdf: float,
        ijk: np.ndarray,
        weighting_function: Optional[Callable[[float], float]] = None,
    ) -> None:
        if weighting_function is not None:
            return self._volume._update_tsdf(sdf, ijk, weighting_function)
        return self._volume._update_tsdf(sdf, ijk)

    def extract_triangle_mesh(self, fill_holes: bool = True, min_weight: float = 0.0) -> Tuple:
        """Returns a the vertices and triangles representing the constructed the TriangleMesh.

        If you can afford to use Open3D as dependency just pass the output of this function to the
        TriangleMesh constructor from Open3d.

        vertices, triangles = integrator.extract_triangle_mesh()
        mesh = o3d.geometry.TriangleMesh(
            o3d.utility.Vector3dVector(vertices),
            o3d.utility.Vector3iVector(triangles),
        )
        """
        vertices, triangles = self._volume._extract_triangle_mesh(fill_holes, min_weight)
        return np.asarray(vertices), np.asarray(triangles)

    def extract_vdb_grids(self, out_file: str) -> None:
        """For now, write the internal map representation to a file.

        Contains both D(x) and W(x) grids.
        """
        self._volume._extract_vdb_grids(out_file)

    def prune(self, min_weight: float):
        """Use the W(x) weights grid to cleanup the generated signed distance field according to a
        minimum weight threshold.

        This function is ideal to cleanup the TSDF grid:D(x) before exporting it.
        """
        return self._volume._prune(min_weight)

class ImplicitRegistration:
    def __init__(
        self,
        vdb_volume_global: VDBVolume,
        max_iters: int,
        convergence_threshold: float,
        clipping_range: float
    ):
        self._registration_pipeline = vdbfusion_pybind._ImplicitRegistration(
            vdb_volume_global=vdb_volume_global._volume,
            max_iters=np.int32(max_iters),
            convergence_threshold=np.float32(convergence_threshold),
            clipping_range=np.float32(clipping_range)
        )

        self.vdb_volume_global = self._registration_pipeline._vdb_volume_global
        self.max_iters = max_iters
        self.convergence_threshold = convergence_threshold
        self.clipping_range = clipping_range

    def __repr__(self) -> str:
        return (
            f"Implicit Registration Pipeline with:\n"
            f"global_vdb_volume = {self.vdb_volume_global.__repr__()}\n"
            f"convergence_threshold    = {self.convergence_threshold}\n"
            f"max_iters     = {self.max_iters}\n"
            f"clipping_range = {self.clipping_range}\n"
        )
    
    def alignScan(self, points: np.ndarray, T_init: np.ndarray) -> Tuple:
        assert isinstance(points, np.ndarray), "points must by np.ndarray(n, 3)"
        assert points.dtype == np.float64, "points dtype must be np.float64"
        assert isinstance(T_init, np.ndarray), "initial guess tf must by np.ndarray"
        assert T_init.dtype == np.float64, "initial guess tf dtype must be np.float64"
        assert T_init.shape == (4, 4) , "initial guess tf must be a (4,4) matrix"
        _points = vdbfusion_pybind._VectorEigen3d(points)
        aligned_points, T, n_iters = self._registration_pipeline._align_scan(_points, T_init)
        return np.asarray(aligned_points), T, n_iters